### PR TITLE
use getentropy() for BSD systems

### PR DESCRIPTION
**Pull request by bayramcicek**
Change-Id: Id6a2629c06d641eb4e7cf3991de4036d2f7b346e

* Resolves: #5851 
* Target version: master 

### Summary

- instead of using /dev/[u]random devices, 
use getentropy() to make direct system calls if the system is a BSD.

- if getentropy() fails, we need to fall back to "/dev/[u]random" approach.

### TODO

- [ ] ...

### Checklist

- [x] Code is properly formatted
- [x] All commits have Change-Id
- [ ] I have run tests with `make check`
- [x] I have issued `make run` and manually verified that everything looks okay
- [x] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by bayramcicek** [comment]
Hi @Ashod. I'd appreciate it if you could review this patch.

File: common/Util.cpp [hunk 1]
```diff
@@ -131,7 +138,13 @@ namespace Util
             // a poor fallback but something.
             std::vector<char> random = getBytes(length);
             int len = 0;
-#ifdef HAVE_SYS_RANDOM_H
+#ifdef BSD
+            len = getentropy(random.data(), length);
+
+            // if getentropy() fails, it will return -1.
+            // then, we should fall back to "/dev/[u]random" approach.
```
Comment (Ashod) [inline common/Util.cpp:145]: I guess we could say `we fall back to getrandom(), if available, otherwise we try "/dev/[u]random".` (or something similar).

**Review by Ashod — APPROVED** [review]
Thanks @bayramcicek! Sorry for the delayed review. Looks good to me, but added @arrowd to get more feedback. Let's wait for Gleb's input, otherwise, happy to merge this soon.

**Comment by arrowd** [comment]
Looks good, although I'd introduce the `HAVE_GETENTROPY` check and use that, rather than using platform checks.

File: common/Util.cpp (hunk 1 above)
Comment (bayramcicek) [reply common/Util.cpp:145]: > I guess we could say `we fall back to getrandom(), if available, otherwise we try "/dev/[u]random".` (or something similar).

But if we are in `#ifdef BSD` code block and  `getentropy()` fails in here, then we can't(?) execute `#elif defined HAVE_SYS_RANDOM_H` code clock - because it is `#elif` directive? Then we can't fall back to `getrandom()`, but we can only use `"/dev/[u]random"` - if `getentropy()` fails. Therefore, imho, we can(?) only use something like `// if getentropy() fails, try "/dev/[u]random"` - if we stick with the current approach.

but:

If we want to fall back to `getrandom()` if `getentropy()` fails in `#ifdef BSD` code block(as you stated), then I can update the code accordingly.

(to sum up) Which one we should use? :
1) if `BSD` -> use `getentropy()` -> if `getentropy()` fails -> (skip `getrandom()`) -> use `"/dev/urandom"`. (current approach)
or
2) if `BSD` -> use `getentropy()` -> if `getentropy()` fails -> use `getrandom()` -> if `getrandom()` fails -> use `"/dev/urandom"`.

Thanks for the review.

Comment (arrowd) [reply common/Util.cpp:145]: According to man, `getentropy` fails either with `EFAULT` due to invalid buffer pointer or `EIO` in case of "some other fatal error", which probably means that there's something really wrong with the system.

And at least of FreeBSD `getrandom` does the same as `getentropy`, so I think there is no point in trying both functions.

**Comment by bayramcicek** [comment]
> Looks good, although I'd introduce the `HAVE_GETENTROPY` check and use that, rather than using platform checks.

Imho;
On my system(Manjaro/Linux), I can use `getentropy` from both `<unistd.h>` and `<sys/random.h>`. So, as far as I understand(but I may be wrong), linux-based systems can also use `getentropy` in `<unistd.h>`. But we should use `getentropy` only for BSD systems. Therefore, I didn't remove the platform check, but moved it to `configure.ac` file.

Thanks for the reviews!

File: configure.ac [hunk 2]
```diff
@@ -1100,6 +1100,31 @@ else
     AC_DEFINE([HAVE_SYS_RANDOM_H],0,[Define to 1 if you have the <sys/random.h> header file.])
 fi
 
+HAVE_GETENTROPY=
+AC_LANG_PUSH([C++])
```
Comment (arrowd) [inline configure.ac:1104]: Why is this needed?

File: configure.ac [hunk 3]
```diff
@@ -1100,6 +1100,31 @@ else
...
+AC_LINK_IFELSE(
+    [AC_LANG_SOURCE([[
+    #if defined(__unix__) || !defined(__APPLE__) && defined(__MACH__)
+        #include <sys/param.h>
+        #if defined(BSD) // FreeBSD, NetBSD, OpenBSD, DragonFly BSD
```
Comment (arrowd) [inline configure.ac:1109]: Why check for BSD? Ultimately we're onle interested if we have `getentropy` in the `unistd.h` header. I'd just check for that without looking back at underlying OS.

**Review by arrowd — APPROVED** [review]

File: configure.ac (hunk 2 above)
Comment (bayramcicek) [reply configure.ac]: I realize that there is no need to check the script for C++ compiler. Man page of autoconf (1) says `configure scripts check for the C compiler and its features by default.`. Then, imho, `AC_LANG_PUSH([C++])` is not necessary so I removed it.

(1): [https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Language-Choice.html](https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Language-Choice.html)

**Comment by bayramcicek** [comment]
Hi @Ashod. Should we merge this patch? 
Thanks.

**Comment by mmeeks** [comment]
@bayramcicek looks like a great patch - can you re-base and get this merged ? =)

File: common/Util.cpp [hunk 4]
```diff
@@ -114,7 +118,12 @@ namespace Util
             std::vector<char> v(length);
 
             int len = 0;
-#ifdef HAVE_SYS_RANDOM_H
+#ifdef HAVE_GETENTROPY
+            len = getentropy(v.data(), length);
+
+            // if getentropy() fails, try "/dev/[u]random"
+            if (len == -1)
```
Comment (thebearon) [inline common/Util.cpp:125]: I find the if condition and body being separated by macro checks (and thus possibly not even being a conditional block) a bit dangerous. Not sure what the best way to handle it would be, I would perhaps add a bool variable for the if condition that has a default value true, and the code inside the #ifdef/#elif sets it according to the condition.

Comment (bayramcicek) [reply common/Util.cpp]: Done.
Thanks.

**Review by mmeeks — APPROVED** [review]
Great - lets get it in =)
